x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/action/TransportPutAutoscalingPolicyActionIT.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.autoscaling.action;
+
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.xpack.autoscaling.AutoscalingIntegTestCase;
+import org.elasticsearch.xpack.autoscaling.AutoscalingMetadata;
+import org.elasticsearch.xpack.autoscaling.policy.AutoscalingPolicy;
+
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.elasticsearch.xpack.autoscaling.AutoscalingTestCase.mutateAutoscailngDeciders;
+import static org.elasticsearch.xpack.autoscaling.AutoscalingTestCase.randomAutoscalingPolicy;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasKey;
+import static org.hamcrest.Matchers.sameInstance;
+
+public class TransportPutAutoscalingPolicyActionIT extends AutoscalingIntegTestCase {
+
+    public void testAddPolicy() {
+        final AutoscalingPolicy policy = putRandomAutoscalingPolicy();
+        final ClusterState state = client().admin().cluster().prepareState().get().getState();
+        final AutoscalingMetadata metadata = state.metaData().custom(AutoscalingMetadata.NAME);
+        assertNotNull(metadata);
+        assertThat(metadata.policies(), hasKey(policy.name()));
+        assertThat(metadata.policies().get(policy.name()).policy(), equalTo(policy));
+    }
+
+    public void testUpdatePolicy() {
+        final AutoscalingPolicy policy = putRandomAutoscalingPolicy();
+        final AutoscalingPolicy updatedPolicy = new AutoscalingPolicy(policy.name(), mutateAutoscailngDeciders(policy.deciders()));
+        putAutoscalingPolicy(updatedPolicy);
+        final ClusterState state = client().admin().cluster().prepareState().get().getState();
+        final AutoscalingMetadata metadata = state.metaData().custom(AutoscalingMetadata.NAME);
+        assertNotNull(metadata);
+        assertThat(metadata.policies(), hasKey(policy.name()));
+        assertThat(metadata.policies().get(policy.name()).policy(), equalTo(updatedPolicy));
+    }
+
+    public void testNoOpPolicy() {
+        final AutoscalingPolicy policy = putRandomAutoscalingPolicy();
+        final ClusterState beforeState = client().admin().cluster().prepareState().get().getState();
+        putAutoscalingPolicy(policy);
+        final ClusterState afterState = client().admin().cluster().prepareState().get().getState();
+        assertThat(beforeState.custom(AutoscalingMetadata.NAME), sameInstance(afterState.custom(AutoscalingMetadata.NAME)));
```

Comment:
I am mildly mystified by how this works (same instance), given that we pick a random node's client in each of the calls and have multiple nodes running? Nevertheless, I ran the test 100 times, so this is purely for my education....

Reply:
Ouch, great catch. This should not be working, and I can tell you why it is, because there's a horrible bug in the assertion! The bug is:

```
assertThat(beforeState.custom(AutoscalingMetadata.NAME), sameInstance(afterState.custom(AutoscalingMetadata.NAME)))
```

should be

```
assertThat(beforeState.metadata().custom(AutoscalingMetadata.NAME), sameInstance(afterState.metadata().custom(AutoscalingMetadata.NAME)))
```

Oops!

Great catch though.

Reply:
I pushed a fix.